### PR TITLE
Support bond response format

### PR DESCRIPTION
**Pull request by feihugis**
Model inference code can fully control the response json schema. The sever code just add these json responses to the return byte stream.

### CODE REVIEW

File: src/http_server.cc [hunk 1]
```diff
@@ -3155,6 +3182,105 @@ HTTPAPIServer::InferRequestClass::FinalizeResponse(
...
+  evbuffer* response_body = response_placeholder;
+  switch (response_compression_type_) {
+    case DataCompressor::Type::DEFLATE:
+    case DataCompressor::Type::GZIP: {
+      auto compressed_buffer = evbuffer_new();
+      auto err = DataCompressor::CompressData(
+          response_compression_type_, response_placeholder, compressed_buffer);
+      if (err == nullptr) {
+        response_body = compressed_buffer;
+        evbuffer_free(response_placeholder);
```
Comment (JiushengChen) [inline src/http_server.cc]: will it be a double free with line 3279?

Comment (JiushengChen) [reply src/http_server.cc]: and response_body will be invalid? response_body is a pointer of response_placeholder

File: src/http_server.cc [hunk 2]
```diff
@@ -3155,6 +3182,105 @@ HTTPAPIServer::InferRequestClass::FinalizeResponse(
...
+      } else {
+        // just log the compression error and return the uncompressed data
+        LOG_VERBOSE(1) << "unable to compress response: "
+                       << TRITONSERVER_ErrorMessage(err);
+        TRITONSERVER_ErrorDelete(err);
+        evbuffer_free(compressed_buffer);
```
Comment (JiushengChen) [inline src/http_server.cc]: if compressed_buffer should be released as well in the other case?

File: src/http_server.cc [hunk 3]
```diff
@@ -3155,6 +3182,105 @@ HTTPAPIServer::InferRequestClass::FinalizeResponse(
...
+    case DataCompressor::Type::IDENTITY:
+    case DataCompressor::Type::UNKNOWN:
+      // Do nothing for other cases
+      break;
+  }
+  SetResponseHeader(!ordered_buffers.empty(), buffer.Size());
+  evbuffer_add_buffer(req_->buffer_out, response_body);
+  // Destroy the evbuffer object as the data has been moved
+  // to HTTP response buffer
+  evbuffer_free(response_body);
```
Comment (JiushengChen) [inline src/http_server.cc]: the mem management of triton classes looks tricky. Please run a mem leak test using valgrind.

File: src/http_server.cc [hunk 4]
```diff
@@ -2813,6 +2813,18 @@ HTTPAPIServer::InferRequestClass::InferRequestClass(
   evhtp_connection_t* htpconn = evhtp_request_get_connection(req);
   thread_ = htpconn->thread;
   evhtp_request_pause(req);
+
+  auto request_type = GetEnvironmentVariableOrDefault("RequestType", "DEFAULT");
```
Comment (JiushengChen) [inline src/http_server.cc]: should be response type? use an env like AB_RESPONSE_TYPE to have the same style of our other env vars.

File: src/http_server.cc (hunk 1 above)
Comment (feihugis) [reply src/http_server.cc]: I did not change the free logic here, which is as same as the http server. 

`response_placeholder` refers to different mem addr (the one before compression) than `response_body` which refers to the compressed one,

File: src/http_server.cc (hunk 2 above)
Comment (feihugis) [reply src/http_server.cc]: It covers all the cases (both DEFLATE and GZIP) with compressions.

File: src/http_server.cc (hunk 3 above)
Comment (feihugis) [reply src/http_server.cc]: Will do it check.

**Review by JiushengChen — APPROVED** [review]

File: src/http_server.cc (hunk 3 above)
Comment (JiushengChen) [reply src/http_server.cc]: no worry, i will check after integrating it with my changes

**Pull request merged**